source/tutorials/learning-journey/sharing-dependencies.md

```diff
@@ -0,0 +1,198 @@
+
+There is a better way!
+
+## Getting started
+
+Create a directory called `shared_project` and enter it:
+```console
+$ mkdir shared_project
+$ cd shared_project
+```
+
+You'll be creating a Python web application as an example project, but don't worry, you'll be given all of the code you need and won't need to know Python to proceed.
+
+Create a file `app.py` with the following contents:
+```python
+from flask import Flask
+
+app = Flask(__name__)
+
+@app.route("/")
+def hello_world():
+    return "<p>Hello, World!</p>"
+```
+This creates a web application that returns `<p>Hello, World!</p>` on the `/` route.
+
+Next create a `pyproject.toml` file with the following contents:
+```toml
+[build-system]
+requires = ["setuptools", "setuptools-scm"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "shared_project"
+version = "0.0.1"
+```
+This file tells Python how to build the project.
+
+For the Nix part of the project you'll create two files: `build.nix` and `default.nix`.
+The actual build recipe will be in `build.nix` and `default.nix` will import this file to perform the build.
+
+First create a `build.nix` file like this:
+```nix
+{
+  python3Packages,
+}:
+
+python3Packages.buildPythonApplication {
+  pname = "shared_project";
+  version = "0.0.1";
+  format = "pyproject";
+  src = builtins.path { path = ./.; name = "shared_project_source"; };
+  propagatedBuildInputs = with python3Packages; [
+    setuptools-scm
+    flask
+  ];
+}
+```
+The Nix expression in this file is a function that takes `python3Packages` as input and produces a derivation.
+This method of defining builds is a common design pattern in the Nix community, and is the format used throughout the `nixpkgs` repository.
+This particular derivation builds your Python application and ensures that `flask`, the library used to create the web application, is available at runtime for the application.
+
+Finally, create a `default.nix` that looks like this:
+```nix
+let
+  pkgs = import <nixpkgs> {};
+in
+  pkgs.callPackage ./build.nix {}
+```
+The `callPackage` function reads the expression in `build.nix` to determine which inputs it needs (in this case, python3Packages), then calls the expression with the inputs that were requested.
+You can read more about `callPackage` in the [Nix Pills][nix_pills_callpackage].
+
+[nix_pills_callpackage]: https://nixos.org/guides/nix-pills/callpackage-design-pattern.html
+
+If you should now be able to build the project with the `nix-build` command.
+If you now execute the `nix-shell` command you should also be put into a shell with the `python3` and `flask` commands available.
+
+## Adding development packages
+As mentioned above, you'll want to add some development packages, but you won't want to add them to `build.nix` or `default.nix`.
+It's time to create a `shell.nix`.
+
+Create the following `shell.nix` file:
```

Comment:
Done with a `default.nix`:
```nix
let
  pkgs = ...;
  build = ...;
in {
  inherit build;
  shell = pkgs.mkShell { ... };
}
```

Then in `shell.nix`:
```nix
(import ./default.nix).shell
```

Then `nix-shell` works.

Then later extend the `default.nix` to have arguments to customize the shell/build:

```nix
{ enableDebugging ? false, optimise ? false }:
...
```

Then change the `shell.nix` to
```nix
(import ./default.nix { }).shell
```

Then one can do `nix-shell --arg enableDebugging true`, or `nix-build --arg optimise true`.

Reply:
That would then require declaring a Nixpkgs import twice:
```nix
# default.nix
let
  pkgs = import <nixpkgs> {};
in {
  foo = pkgs.callPackage ./foo.nix {};
}
```

```nix
# shell.nix
let
  pkgs = import <nixpkgs> {};
  build = import ./default.nix;
in pkgs.mkShell {
  inputsFrom = [ build ];
  packages = [ ... ];
}
```

Which then causes problems when you decide to want to update Nixpkgs, because now you need to update it in two places. This could be fixed by having a `common.nix` file containing all the common expressions needed for both the build and the shell, but that's getting a bit complicated then.

Reply:
Having thought about this for a bit I still think having `default.nix` be a function is outside the scope of the tutorial, especially given that in this case it only really takes `pkgs` as an argument. It may be a better pattern in the long run, but it feels unnecessary when the "project" being worked on in the tutorial is ~10 lines of Python.

I wouldn't be opposed to creating a "parameterized builds" tutorial/recipe. It probably works best as a recipe given that you can really quickly demonstrate how to make it work and what the benefits are.